### PR TITLE
Safe zones for robot

### PR DESCRIPTION
**Pull request by Christdej**

### CODE REVIEW

File: backend/api/Controllers/RobotController.cs [hunk 1]
```diff
@@ -654,6 +657,99 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
 
+
+    /// <summary>
+    /// Starts a mission which drives the robot to the nearest safe position
+    /// </summary>
+    /// <remarks>
+    /// <para> This query starts a localization for a given robot </para>
+    /// </remarks>
+    [HttpPost]
+    [Route("{robotId}/{asset}/{deck}/safe-position")]
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:668]:
```suggestion
    [Route("{robotId}/{asset}/{deck}/go-to-safe-position")]
```
I think the endpoint name should reflect the action it is provoking

File: backend/api/Controllers/RobotController.cs [hunk 2]
```diff
@@ -654,6 +657,99 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        {
+            _logger.LogWarning("Could not find robot with id={id}", robotId);
+            return NotFound("Robot not found");
+        }
+
+        var assetDeck = await _assetDeckService.ReadByAssetAndDeck(asset, deck);
+        if (assetDeck is null)
+        {
+            _logger.LogWarning("Could not find any safe positions for asset={asset} and deck={deck}", asset, deck);
+            return NotFound("No safe position found");
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:693]: I think this should be split in 3 checks, one for asset, then for deck, then for safe positions

File: backend/api/Controllers/RobotController.cs [hunk 3]
```diff
@@ -654,6 +657,99 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+            RobotPose = closestSafePosition,
+            Inspections = new List<CustomInspectionQuery>(),
+            InspectionTarget = new Position(),
+            TaskOrder = 0
+        };
+        var mission = new Mission
+        {
+            Name = "Drive to Safe Position Mission",
+            Robot = robot,
+            AssetCode = "NA",
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:709]:
```suggestion
            AssetCode = "assetDeck.AssetCode",
```

File: backend/api/Controllers/RobotController.cs [hunk 4]
```diff
@@ -654,6 +657,99 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        var customTaskQuery = new CustomTaskQuery
+        {
+            RobotPose = closestSafePosition,
+            Inspections = new List<CustomInspectionQuery>(),
+            InspectionTarget = new Position(),
+            TaskOrder = 0
+        };
+        var mission = new Mission
+        {
+            Name = "Drive to Safe Position Mission",
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:707]:
```suggestion
            Name = "Drive to Safe Position",
```

File: backend/api/Controllers/RobotController.cs [hunk 5]
```diff
@@ -654,6 +657,99 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+            Robot = robot,
+            AssetCode = "NA",
+            EchoMissionId = 0,
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:710]:
```suggestion
            EchoMissionId = null,
```

File: backend/api/Controllers/RobotController.cs [hunk 6]
```diff
@@ -654,6 +657,99 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        catch (HttpRequestException e)
+        {
+            string message = $"Could not reach ISAR at {robot.IsarUri}";
+            _logger.LogError(e, "{message}", message);
+            OnIsarUnavailable(robot);
+            return StatusCode(StatusCodes.Status502BadGateway, message);
+        }
+        catch (MissionException e)
+        {
+            _logger.LogError(e, "Error while starting ISAR localization mission");
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:731]:
```suggestion
            _logger.LogError(e, "Error while starting ISAR mission");
```

File: backend/api/Controllers/RobotController.cs [hunk 7]
```diff
@@ -674,4 +770,34 @@ private async void OnIsarUnavailable(Robot robot)
         robot.CurrentMissionId = null;
         await _robotService.Update(robot);
     }
+
+    private static Pose ClosestSafePosition(Pose robotPose, IList<SafePosition> safePositions)
+    {
+        if (safePositions == null || safePositions.Count == 0)
+        {
+            throw new ArgumentException("Poses list cannot be null or empty.");
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:778]:
```suggestion
            throw new ArgumentException("List of safe positions cannot be null or empty.");
```

File: backend/api/Controllers/RobotController.cs [hunk 8]
```diff
@@ -674,4 +770,34 @@ private async void OnIsarUnavailable(Robot robot)
...
+        }
+        return closestPose;
+    }
+
+    private static float CalculateDistance(Pose pose1, Pose pose2)
+    {
+        var pos1 = pose1.Position;
+        var pos2 = pose2.Position;
+        return (float)Math.Sqrt(Math.Pow(pos1.X - pos2.X, 2) + Math.Pow(pos1.Y - pos2.Y, 2) + Math.Pow(pos1.Z - pos2.Z, 2));
+    }
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:801]: Maybe we should create an issue to calculate closest safe position based on cost map and A*

File: backend/api/Services/AssetDeckService.cs [hunk 9]
```diff
@@ -11,10 +11,16 @@ public interface IAssetDeckService
 
         public abstract Task<AssetDeck?> ReadById(string id);
 
+        public abstract Task<AssetDeck?> ReadByAssetAndDeck(string asset, string deck);
+
         public abstract Task<AssetDeck> Create(CreateAssetDeckQuery newAssetDeck);
 
+        public abstract Task<AssetDeck> Create(CreateAssetDeckQuery newAssetDeck, Pose safePosition);
```
Comment (GodVenn) [inline backend/api/Services/AssetDeckService.cs:18]:
```suggestion
        public abstract Task<AssetDeck> Create(CreateAssetDeckQuery newAssetDeck, List<Pose> safePositions);
```
If we should have this method, shouldn't it have a list as input?

**Review by GodVenn — COMMENTED** [review]
In the API call to go to safe position, we should first cancel active missions on the robot, so that it goes to the safe position immediately

File: backend/api/Controllers/RobotController.cs (hunk 1 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:668]: Agreed

File: backend/api/Controllers/RobotController.cs (hunk 3 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:709]: Agreed

File: backend/api/Controllers/RobotController.cs (hunk 5 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:710]: EchoMissionId is an int, so cannot be null

File: backend/api/Controllers/RobotController.cs (hunk 4 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:707]: Agreed

File: backend/api/Controllers/AssetDeckController.cs [hunk 10]
```diff
@@ -105,6 +105,57 @@ public async Task<ActionResult<AssetDeck>> Create([FromBody] CreateAssetDeckQuer
...
+    /// <para> This query adds a new safe position to the database </para>
+    /// </remarks>
+    [HttpPost]
+    [Route("{asset}/{deck}/safe-position")]
+    [ProducesResponseType(typeof(AssetDeck), StatusCodes.Status201Created)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+    public async Task<ActionResult<AssetDeck>> AddSafePosition(
```
[the remark this answers is not in the document]
Comment (Christdej) [reply backend/api/Controllers/AssetDeckController.cs:122]: Agreed

File: backend/api/Controllers/RobotController.cs (hunk 2 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:693]: Do you think this is necessary? Same message and status code for all the checks, just more code for the same outcome?

File: backend/api/Controllers/RobotController.cs (hunk 6 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:731]: Agreed

File: backend/api/Controllers/RobotController.cs (hunk 8 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:801]: Requires that we have the cost map available for all robots?

File: backend/api/Services/AssetDeckService.cs (hunk 9 above)
Comment (Christdej) [reply backend/api/Services/AssetDeckService.cs:18]: Agreed

File: backend/api/Controllers/RobotController.cs (hunk 2 above)
Comment (GodVenn) [reply backend/api/Controllers/RobotController.cs:693]: Not the same message, "Asset not found", "Deck not found" and "No safe positions on this asset deck"

File: backend/api/Controllers/RobotController.cs (hunk 5 above)
Comment (GodVenn) [reply backend/api/Controllers/RobotController.cs:710]: Ah, should be nullable, especially now that we have custom missions

File: backend/api/Controllers/RobotController.cs (hunk 8 above)
Comment (GodVenn) [reply backend/api/Controllers/RobotController.cs:801]: Good point

File: backend/api/Controllers/RobotController.cs (hunk 5 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:745]: Made an issue

https://github.com/equinor/flotilla/issues/794

File: backend/api/Controllers/AssetDeckController.cs [hunk 11]
```diff
@@ -105,6 +112,58 @@ public async Task<ActionResult<AssetDeck>> Create([FromBody] CreateAssetDeckQuer
...
+        [FromBody] Pose safePosition
+    )
+    {
+        _logger.LogInformation("Adding new safe position");
+        try
+        {
+            var assetDeck = await _assetDeckService.AddSafePosition(asset, deck, new SafePosition(safePosition));
+            if (assetDeck != null)
+            {
+                _logger.LogInformation("Succesfully added new safe position for asset '{assetId}' and deck '{deckId}', creating new", asset, deck);
```
Comment (GodVenn) [inline backend/api/Controllers/AssetDeckController.cs:142]:
```suggestion
                _logger.LogInformation("Succesfully added new safe position for asset '{assetId}' and deck '{deckId}'", asset, deck);
```

File: backend/api/Controllers/RobotController.cs [hunk 12]
```diff
@@ -654,6 +657,134 @@ public async Task<ActionResult> ResumeMission([FromRoute] string robotId)
...
+        var robot = await _robotService.ReadById(robotId);
+        if (robot == null)
+        {
+            _logger.LogWarning("Could not find robot with id={id}", robotId);
+            return NotFound("Robot not found");
+        }
+
+        var assets = await _assetDeckService.ReadByAsset(asset);
+
+        if (assets.Any())
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:691]:
```suggestion
        if (!assets.Any())
```

File: backend/api/Controllers/RobotController.cs [hunk 13]
```diff
@@ -674,4 +805,34 @@ private async void OnIsarUnavailable(Robot robot)
         robot.CurrentMissionId = null;
         await _robotService.Update(robot);
     }
+
+    private static Pose ClosestSafePosition(Pose robotPose, IList<SafePosition> safePositions)
+    {
+        if (safePositions == null || safePositions.Any())
```
Comment (GodVenn) [inline backend/api/Controllers/RobotController.cs:811]:
```suggestion
        if (safePositions == null || !safePositions.Any())
```

File: backend/api/Services/AssetDeckService.cs [hunk 14]
```diff
@@ -49,20 +57,58 @@ private IQueryable<AssetDeck> GetAssetDecks()
                 .FirstOrDefaultAsync(assetDeck => assetDeck.Id.Equals(id));
         }
 
-        public async Task<AssetDeck> Create(CreateAssetDeckQuery newAssetDeck)
+        public async Task<IEnumerable<AssetDeck>> ReadByAsset(string asset)
+        {
+            return await _context.AssetDecks.Where(a =>
+                a.AssetCode.Equals(asset)).Include(a => a.SafePositions).ToListAsync();
```
Comment (GodVenn) [inline backend/api/Services/AssetDeckService.cs:63]:
```suggestion
                a.AssetCode.ToLower().Equals(asset.ToLower())).Include(a => a.SafePositions).ToListAsync();
```
Asset code string comparison should be case insensitive

File: backend/api/Services/AssetDeckService.cs [hunk 15]
```diff
@@ -49,20 +57,58 @@ private IQueryable<AssetDeck> GetAssetDecks()
...
+        }
+
+        public async Task<AssetDeck?> ReadByAssetAndDeck(string asset, string deck)
+        {
+            return await _context.AssetDecks.Where(a =>
+                a.AssetCode.Equals(asset) &&
+                a.DeckName.Equals(deck)
```
Comment (GodVenn) [inline backend/api/Services/AssetDeckService.cs:70]: Asset code and deck name string comparison should be case insensitive

File: backend/api/Services/AssetDeckService.cs [hunk 16]
```diff
@@ -49,20 +57,58 @@ private IQueryable<AssetDeck> GetAssetDecks()
...
+        }
+
+        public async Task<AssetDeck?> AddSafePosition(string asset, string deck, SafePosition safePosition)
+        {
+            var assetDeck = await ReadByAssetAndDeck(asset, deck);
+            if (assetDeck is null)
+            {
+                return null;
+            }
+            assetDeck.SafePositions.Add(safePosition);
```
Comment (GodVenn) [inline backend/api/Services/AssetDeckService.cs:106]: Have you tested this if the asset deck has no list to begin with? 
I think you might have to initialize the list in that case

File: backend/api/Controllers/AssetDeckController.cs (hunk 11 above)
Comment (Christdej) [reply backend/api/Controllers/AssetDeckController.cs:142]: Fixed

File: backend/api/Controllers/RobotController.cs (hunk 12 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:691]: Good catch

File: backend/api/Controllers/RobotController.cs (hunk 13 above)
Comment (Christdej) [reply backend/api/Controllers/RobotController.cs:811]: Good catch

File: backend/api/Services/AssetDeckService.cs (hunk 14 above)
Comment (Christdej) [reply backend/api/Services/AssetDeckService.cs:63]: Agreed

File: backend/api/Services/AssetDeckService.cs (hunk 15 above)
Comment (Christdej) [reply backend/api/Services/AssetDeckService.cs:70]: Agreed

File: backend/api/Services/AssetDeckService.cs (hunk 16 above)
Comment (Christdej) [reply backend/api/Services/AssetDeckService.cs:106]: In the initialization of an AssetDeck, an empty list is made, so you are able to use this service even if there are no SafePositions prior

Comment (GodVenn) [reply backend/api/Services/AssetDeckService.cs:108]: Have you tested it? Might behave differently than expected due to C# -> Database -> C#

Comment (Christdej) [reply backend/api/Services/AssetDeckService.cs:108]: Tested it locally, with SQLite.

**Review by GodVenn — COMMENTED** [review]
I see that the SafePositions are created as a table in the SQL database, we should add a "DbSet" in the top of our `FlotillaDbContext` file to reflect this in code and give us an access way to them

**Comment by Christdej** [comment]
> I see that the SafePositions are created as a table in the SQL database, we should add a "DbSet" in the top of our `FlotillaDbContext` file to reflect this in code and give us an access way to them

Agreed

**Review by GodVenn — APPROVED** [review]
LGTM

**Pull request merged**